### PR TITLE
CRUD companies and add ransack search

### PR DESCRIPTION
**Pull request by IvanDanyliv**
[Trello Task](https://trello.com/_____)

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
- [x] screenshot of feature is added if task has UI

image

### CODE REVIEW

File: app/controllers/companies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  def show; end
+
+  def create
+    @company = Company.new(company_params)
+    if @company.save
+      flash[:success] = "Company '#{@company.name}' created."
+      redirect_to @company
+    else
+      render 'new', warning: 'Invalid company parameters'
```
Comment (mmarusyk) [inline app/controllers/companies_controller.rb:21]:
```suggestion
      flash.now[:warning] = 'Invalid company parameters!'
      render :now
```

File: app/javascript/stylesheets/companies.scss [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  }                                                 
+  ul {
+    padding: 0;
+    li {
+      margin: 5px 5px;
+      a {
+        width: 150px;
+      }
+    }
+  }
```
Comment (mmarusyk) [inline app/javascript/stylesheets/companies.scss:47]: add new line at end file

File: app/views/companies/edit.html.slim [hunk 3]
```diff
@@ -0,0 +1,6 @@
+= provide(:button_text, 'Edit company')
+= provide(:cancel_btn_text, 'Cancel')
+= provide(:path, company_path)
+h1 Update company profile
```
Comment (mmarusyk) [inline app/views/companies/edit.html.slim:4]: Enter

File: app/views/companies/new.html.slim [hunk 4]
```diff
@@ -0,0 +1,6 @@
+= provide(:button_text, 'Create new company')
+= provide(:cancel_btn_text, 'Cancel')
+= provide(:path, companies_path)
+h1.title New company
```
Comment (mmarusyk) [inline app/views/companies/new.html.slim:4]: Enter

File: app/javascript/stylesheets/companies.scss [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      width: 240px;
+      font-size: 24px;
+      display: inline-flex;
+    }
+    .value {
+      width: 340px;
+    }
+  span {
+    margin-right: 10px;
+  }
```
Comment (mmarusyk) [inline app/javascript/stylesheets/companies.scss:35]: Align code

File: spec/controllers/companies_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  describe 'GET#show' do
+    before do
+      get :show, params: { id: company.id }
+    end
+
+    it 'returns success and assigns company' do
+      expect(response).to have_http_status(:success)
+      expect(assigns(:company)).to eq(company)
+    end
```
Comment (mmarusyk) [inline spec/controllers/companies_controller_spec.rb:25]: You have only one test. Why do you use get :show in before block?

File: app/javascript/stylesheets/companies.scss [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  margin-left: 10px;
+}
+
+.company-info {
+  display: flex;
+  justify-content: flex-start;
+  flex-direction: column;
+  width: 560px;
+}
+.info-item {
```
Comment (Ukhanskyi) [inline app/javascript/stylesheets/companies.scss:22]: mb add enter

File: app/javascript/stylesheets/companies.scss [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  margin: 5px 0 5px 10px;
+}
+.name {
```
Comment (Ukhanskyi) [inline app/javascript/stylesheets/companies.scss:25]: and here

File: app/javascript/stylesheets/companies.scss [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  width: 240px;
+  font-size: 24px;
+  display: inline-flex;
+}
+.value {
```
Comment (Ukhanskyi) [inline app/javascript/stylesheets/companies.scss:30]: and here

File: app/javascript/stylesheets/companies.scss [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  width: 340px;
+}
+span {
```
Comment (Ukhanskyi) [inline app/javascript/stylesheets/companies.scss:33]: here to

File: app/javascript/stylesheets/companies.scss [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  margin-right: 10px;
+}
+nav {
```
Comment (Ukhanskyi) [inline app/javascript/stylesheets/companies.scss:36]: too

File: app/views/companies/edit.html.slim [hunk 12]
```diff
@@ -0,0 +1,8 @@
+= provide(:button_text, 'Edit company')
+= provide(:cancel_btn_text, 'Cancel')
+= provide(:path, company_path)
+
+h1 Update company profile
+.container
+  = render 'form', object: @company
+
```
Comment (Ukhanskyi) [inline app/views/companies/edit.html.slim:8]: mb whitespace here

**Review by Ukhanskyi — APPROVED** [review]

File: app/controllers/companies_controller.rb [hunk 13]
```diff
@@ -0,0 +1,59 @@
+class CompaniesController < ApplicationController
+  before_action :company, only: %i[show edit update destroy]
```
Comment (vasya539) [inline app/controllers/companies_controller.rb:2]: rename to set_company

File: app/controllers/companies_controller.rb [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+  def index
+    @search = Company.search(params[:q])
```
Comment (vasya539) [inline app/controllers/companies_controller.rb:5]: image

File: app/controllers/companies_controller.rb [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    @company.destroy
+    redirect_to [:companies]
+    flash[:danger] = "Company profile \"#{@company.name}\" with id:#{@company.id} has been deleted"
+  end
+
+  private
+
+  def company_params
+    params.require(:company).permit(
+      :name,
```
Comment (vasya539) [inline app/controllers/companies_controller.rb:47]: one new line would be enough

File: app/controllers/companies_controller.rb [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+      :company_type,
+      :phone,
+      :email,
+      :website,
+      :fax
+    )
+  end
+
+  def company
+    @company ||= Company.find(params[:id])
```
Comment (vasya539) [inline app/controllers/companies_controller.rb:57]: why do you have "||"?

File: app/javascript/stylesheets/companies.scss [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+nav {    
+  margin-left: -15px;
+} 
+                                                
+ul {
+  padding: 0;
+  li {
+    margin: 5px 5px;
```
Comment (pereslop) [inline app/javascript/stylesheets/companies.scss:47]: I think margin: 5px is enough

File: app/views/companies/_form.html.slim [hunk 18]
```diff
@@ -0,0 +1,34 @@
+.company-form
```
Comment (pereslop) [inline app/views/companies/_form.html.slim:1]: Why do you need this block, there no content in it

File: app/views/companies/new.html.slim [hunk 19]
```diff
@@ -0,0 +1,7 @@
+= provide(:button_text, 'Create new company')
+= provide(:cancel_btn_text, 'Cancel')
+= provide(:path, companies_path)
+
+h1.title New company
+.container
```
Comment (pereslop) [inline app/views/companies/new.html.slim:6]: Why you add here .container class? Do you want to use bootsttap grid here? If answer is yes, also should be .row, .col blocks

File: app/models/company.rb [hunk 20]
```diff
@@ -29,7 +29,7 @@ class Company < ApplicationRecord
     OTHER
   ].freeze
 
-  enum company_type: COMPANY_TYPES
+  enum company_types: COMPANY_TYPES
```
Comment (chu4) [inline app/models/company.rb:32]: should be `company_type`
problem with rebasing or smth

File: app/controllers/companies_controller.rb (hunk 14 above)
Comment (pereslop) [reply app/controllers/companies_controller.rb]: one more, please add pagination to companies, it is bad if you will load all companies on the page

File: app/views/companies/_company.html.slim [hunk 21]
```diff
@@ -0,0 +1,23 @@
+.col-md-4
+  p
+    | Company name:
```
Comment (pereslop) [inline app/views/companies/_company.html.slim:3]: add I18n

File: app/views/companies/index.html.slim [hunk 22]
```diff
@@ -0,0 +1,11 @@
+.container-wrapper
+  .new-button
+    = link_to 'New company',
+      new_company_path,
+      class: 'btn btn-primary'
+  h1
+    | All companies
```
Comment (pereslop) [inline app/views/companies/index.html.slim:7]: I18n

File: app/views/companies/show.html.slim [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+      span.name Company type: 
+      span.value = @company.company_type
+    .info-item
+      span.name Phone: 
+      span.value = @company.phone
+    .info-item
+      span.name Email: 
+      span.value = @company.email
+    .info-item
+      span.name Website:
```
Comment (pereslop) [inline app/views/companies/show.html.slim:18]: I18n

File: app/javascript/stylesheets/application.scss [hunk 24]
```diff
@@ -1,7 +1,102 @@
...
+
+.navbar-header{
+  background-color: rgba(255, 255, 255, 0.15);
+}
+
+.nav-links {
+  color: white;
+}
+
+.nav-links:hover{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:45]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 25]
```diff
@@ -1,7 +1,102 @@
...
+  color: whitesmoke;
+  background-color: #556588;
+}
+
+#dropdownMenuButton{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:50]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 26]
```diff
@@ -1,7 +1,102 @@
...
+  width: 158px;
+}
+
+.dropdown-lang {
+  background: #798db6;
+  width: 155px;
+}
+
+.footer{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:59]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 27]
```diff
@@ -1,7 +1,102 @@
...
+  background: #556588;
+  color:whitesmoke;
+  
+  .links{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:63]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 28]
```diff
@@ -1,7 +1,102 @@
...
+    ul {list-style-type: none;}
+    li a{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:65]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 29]
```diff
@@ -1,7 +1,102 @@
...
+      color: whitesmoke;
+      transition: color .2s;
+      &:hover{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:68]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 30]
```diff
@@ -1,7 +1,102 @@
...
+        text-decoration:none;
+        color:#518fda;
+        }
+    }
+  } 
+
+  .about-company{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:75]: add space before bracket

File: app/javascript/stylesheets/application.scss [hunk 31]
```diff
@@ -1,7 +1,102 @@
...
+    i{font-size: 25px;}
+    a{
```
Comment (pereslop) [inline app/javascript/stylesheets/application.scss:77]: add space before bracket

File: app/models/company.rb [hunk 32]
```diff
@@ -31,8 +31,6 @@ class Company < ApplicationRecord
 
   enum company_type: COMPANY_TYPES
 
-  before_save { email.downcase! }
```
Comment (pereslop) [inline app/models/company.rb:34]: do you really need to remove this?

Comment (pereslop) [reply app/models/company.rb:34]: if you handle this in other place, it is OK

File: app/views/account/admin/companies/_company.html.slim [hunk 33]
```diff
@@ -0,0 +1,24 @@
+.col-md-4
+  p
+    | Company name:
```
Comment (pereslop) [inline app/views/account/admin/companies/_company.html.slim:3]: I18n

File: app/views/account/admin/companies/_form.html.slim [hunk 34]
```diff
@@ -0,0 +1,34 @@
+= simple_form_for [:account, :admin, @company] do |f|
+  .form-group.company-form
+    = f.input :name, label: t('companies.form.name_label'),
+                     placeholder: "#{t('companies.form.name_label')}",
```
Comment (pereslop) [inline app/views/account/admin/companies/_form.html.slim:4]: you don't need interpolation here, and bellow 
`placeholder: t('companies.form.name_label'),` should work 
and in other files too...

File: app/views/account/admin/companies/_search_form.html.slim [hunk 35]
```diff
@@ -0,0 +1,12 @@
+= search_form_for @ransack, url: account_admin_companies_path, builder: SimpleForm::FormBuilder do |f|
+  .form-group.company-form
+    = f.text_field :name_or_email_or_phone_cont, label: 'Name', label: false,
+                                                 placeholder: "#{t('companies.search_form.name_email_phone_placeholder')}",
```
Comment (pereslop) [inline app/views/account/admin/companies/_search_form.html.slim:4]: interpolation

File: app/views/account/admin/companies/edit.html.slim [hunk 36]
```diff
@@ -0,0 +1,6 @@
+= provide(:button_text, t('companies.edit.btn_edit'))
```
Comment (pereslop) [inline app/views/account/admin/companies/edit.html.slim:1]: like here, translation whithout interpolation. good

File: app/views/account/admin/companies/edit.html.slim [hunk 37]
```diff
@@ -0,0 +1,6 @@
...
+= provide(:cancel_btn_text, t('companies.edit.btn_cancel'))
+= provide(:path, account_admin_company_path)
+
+h1 #{t('companies.edit.edit_tittle')}
```
Comment (pereslop) [inline app/views/account/admin/companies/edit.html.slim:5]: h1 = t('companies.......

File: app/views/account/admin/companies/index.html.slim [hunk 38]
```diff
@@ -0,0 +1,9 @@
+.container-wrapper
+  .new-button
+    = link_to new_account_admin_company_path, class: 'btn btn-primary'
+      = t('companies.index.btn_new_company')
+  h1 #{t('companies.index.companies_title')}
```
Comment (pereslop) [inline app/views/account/admin/companies/index.html.slim:5]: h1 = t('companies.....

File: app/views/account/admin/companies/index.html.slim [hunk 39]
```diff
@@ -0,0 +1,9 @@
...
+  =render 'search_form'
+  .row.mt-4
+    - @companies.each do |company|
+      = render company
```
Comment (pereslop) [inline app/views/account/admin/companies/index.html.slim:9]: please add pagination in this PR, or later, don't forget about it. Look at users page

File: app/views/account/admin/companies/new.html.slim [hunk 40]
```diff
@@ -0,0 +1,6 @@
+= provide(:button_text, t('companies.new.btn_create'))
+= provide(:cancel_btn_text, t('companies.new.btn_cancel'))
+= provide(:path, account_admin_companies_path)
+
+h1.title #{t('companies.new.new_tittle')}
```
Comment (pereslop) [inline app/views/account/admin/companies/new.html.slim:5]: h1.title= t('companies....

File: app/views/account/admin/companies/show.html.slim [hunk 41]
```diff
@@ -0,0 +1,29 @@
+.container 
+  h1.title #{t('companies.show.show_title')}
```
Comment (pereslop) [inline app/views/account/admin/companies/show.html.slim:2]: h1.title= ......

File: app/views/account/admin/companies/show.html.slim [hunk 42]
```diff
@@ -0,0 +1,29 @@
+.container 
+  h1.title #{t('companies.show.show_title')}      
+  .company-info      
+    .info-item
+      span.name #{t('companies.show.company_name')}  
+      span.value = @company.name
+    .info-item
+      span.name #{t('companies.show.company_type')}
```
Comment (pereslop) [inline app/views/account/admin/companies/show.html.slim:8]: span.name= t('.....

File: app/views/account/admin/companies/show.html.slim [hunk 43]
```diff
@@ -0,0 +1,29 @@
...
+    .info-item
+      span.name #{t('companies.show.website')} 
+      span.value = @company.website
+    .info-item
+      span.name #{t('companies.show.fax')} 
+      span.value = @company.fax
+    nav
+      ul.nav.navbar
+        li.nav-item = link_to edit_account_admin_company_path, class: "btn btn-outline-warning"
+                      = t('companies.show.btn_edit_company')
```
Comment (pereslop) [inline app/views/account/admin/companies/show.html.slim:25]: good

File: app/views/account/admin/companies/show.html.slim [hunk 44]
```diff
@@ -0,0 +1,29 @@
...
+      span.name #{t('companies.show.phone')} 
+      span.value = @company.phone
+    .info-item
+      span.name #{t('companies.show.email')} 
+      span.value = @company.email
+    .info-item
+      span.name #{t('companies.show.website')} 
+      span.value = @company.website
+    .info-item
+      span.name #{t('companies.show.fax')}
```
Comment (pereslop) [inline app/views/account/admin/companies/show.html.slim:20]: bad

File: db/migrate/20201015161629_change_type_company_type.rb [hunk 45]
```diff
@@ -0,0 +1,5 @@
+class ChangeTypeCompanyType < ActiveRecord::Migration[6.0]
+  def change
+    change_column :companies, :company_type, 'integer USING CAST(company_type AS integer)'
```
Comment (pereslop) [inline db/migrate/20201015161629_change_type_company_type.rb:3]: can you explain what  do`'integer USING CAST(company_type AS integer)'`?

**Review by pereslop — CHANGES_REQUESTED** [review]

File: db/migrate/20201015161629_change_type_company_type.rb (hunk 45 above)
Comment (IvanDanyliv) [reply db/migrate/20201015161629_change_type_company_type.rb:3]: change type company_type from string to integer

File: app/models/company.rb (hunk 32 above)
Comment (chu4) [reply app/models/company.rb:34]: seems weird doesn't look like you changes make sure you dont have some problems with rebasing

**Comment by chu4** [comment]
as there is already an account namespace you can already move your crud there

File: app/controllers/account/admin/companies_controller.rb [hunk 46]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def new
+    @company = Company.new
+  end
+
+  def show; end
+
+  def create
+    @company = Company.new(company_params)
```
Comment (BogdanBusko) [inline app/controllers/account/admin/companies_controller.rb:16]: enter after

File: app/controllers/account/admin/companies_controller.rb [hunk 47]
```diff
@@ -0,0 +1,52 @@
...
+    if @company.save
+      flash[:success] = "Company \"#{@company.name}\" - created with id:#{@company.id}"
```
Comment (BogdanBusko) [inline app/controllers/account/admin/companies_controller.rb:18]: i18n

File: app/javascript/stylesheets/application.scss [hunk 48]
```diff
@@ -1,7 +1,102 @@
...
+      transition: color .2s;
+      &:hover {
+        text-decoration:none;
+        color:#518fda;
+        }
+    }
+  } 
+
+  .about-company {
+    i { font-size: 25px; }
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:76]: new line and between i and a styles enter

File: app/javascript/stylesheets/application.scss [hunk 49]
```diff
@@ -1,7 +1,102 @@
...
+
+.footer {
+  background: #556588;
+  color:whitesmoke;
+  
+  .links {
+    ul { list-style-type: none; }
+    li a{
+      color: whitesmoke;
+      transition: color .2s;
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:67]: enter after

File: app/javascript/stylesheets/application.scss [hunk 50]
```diff
@@ -1,7 +1,102 @@
...
 @import 'users';
 @import 'basic';
-@import 'account';
+@import 'account';  
+
+body {
+  background-color: #e1ebee;
+}
+
+.header {
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:14]: all styles that are used for header move to file header.scss

File: app/javascript/stylesheets/application.scss [hunk 51]
```diff
@@ -1,7 +1,102 @@
...
+  // background: #0070b3;
+  background-image: url("../../../public/ivano-frankivsk2.jpg");
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:16]: create folder assets and add there folder images

Comment (BogdanBusko) [reply app/javascript/stylesheets/application.scss:16]: move this image there

Comment (BogdanBusko) [reply app/javascript/stylesheets/application.scss:16]: also move folder stylesheets to this folder

File: app/javascript/stylesheets/application.scss [hunk 52]
```diff
@@ -1,7 +1,102 @@
...
+    i { font-size: 18px; }
+  }
+
+  .copyright p {
+    border-top:1.5px solid rgba(255, 255, 255, 0.418);
+    font-size: 20px;
+  } 
+}
+
+p {
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:94]: is it to set global styles for p tag?

File: app/javascript/stylesheets/application.scss [hunk 53]
```diff
@@ -1,7 +1,102 @@
...
+#dropdownMenuButton {
+  width: 158px;
+}
+
+.dropdown-lang {
+  background: #798db6;
+  width: 155px;
+}
+
+.footer {
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/application.scss:59]: also fotter to fotter.scss

File: app/javascript/stylesheets/companies.scss [hunk 54]
```diff
@@ -0,0 +1,51 @@
...
+span {
+  margin-right: 10px;
+}
+
+nav {    
+  margin-left: -15px;
+} 
+                                                
+ul {
+  padding: 0;
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/companies.scss:44]: enter after

File: app/javascript/stylesheets/companies.scss [hunk 55]
```diff
@@ -0,0 +1,51 @@
...
+  li {
+    margin: 5px;
```
Comment (BogdanBusko) [inline app/javascript/stylesheets/companies.scss:46]: enter after

File: spec/controllers/account/admin/companies_controller_spec.rb [hunk 56]
```diff
@@ -0,0 +1,123 @@
+require 'rails_helper'
+
+RSpec.describe Account::Admin::CompaniesController, type: :controller do
```
Comment (BogdanBusko) [inline spec/controllers/account/admin/companies_controller_spec.rb:3]: write login_user after this line and remove in all other places

File: spec/controllers/account/admin/companies_controller_spec.rb [hunk 57]
```diff
@@ -0,0 +1,123 @@
...
+      get :edit, params: { id: company.id }
+    end
+
+    it 'returns http success and assign company' do
+      expect(response).to have_http_status(:success)
+      expect(assigns(:company)).to eq(company)
+    end
+  end
+
+  describe 'PUT#update' do\
```
Comment (BogdanBusko) [inline spec/controllers/account/admin/companies_controller_spec.rb:76]: remove \ at the end

File: spec/controllers/account/admin/companies_controller_spec.rb [hunk 58]
```diff
@@ -0,0 +1,123 @@
...
+    login_user
+    context 'with valid params' do
+      before do
+        put :update, params: { id: company.id,
+                               company: valid_params.merge!(name: 'Example',
+                                                            phone: '0123456789',
+                                                            email: 'example@email.com') }
+      end
+
+      it 'assigns the company' do
```
Comment (BogdanBusko) [inline spec/controllers/account/admin/companies_controller_spec.rb:86]: it { expect(assigns(:company)).to eq(company)
it { expect(response).to have_http_status(:redirect) }
it { expect(response).to redirect_to.....
EVERY WERE

File: app/controllers/account/admin/companies_controller.rb [hunk 59]
```diff
@@ -0,0 +1,53 @@
+class Account::Admin::CompaniesController < Account::Admin::AdminController
+  before_action :set_company, only: %i[show edit update destroy]
+
+  def index
+    @ransack = Company.ransack(params[:q])
+    @companies = @ransack.result(distinct: true).page(params[:page]).per(12)
```
Comment (marikrebega) [inline app/controllers/account/admin/companies_controller.rb:6]: if possible then remove distinct

**Review by marikrebega — CHANGES_REQUESTED** [review]
resolve all remarks

File: db/migrate/20201015161629_change_type_company_type.rb (hunk 45 above)
Comment (vasya539) [reply db/migrate/20201015161629_change_type_company_type.rb:3]: > change type company_type from string to integer

image

**Review by mmarusyk — APPROVED** [review]

**Review by Andrii-Kozak — APPROVED** [review]

**Review by chu4 — APPROVED** [review]

**Pull request merged**